File: calculator/src/Cost.ts
import {CratePerQueue, Item, ResourcePerCrate} from "./Models";

export interface Cost {
    bmat: number,
    rmat: number,
    emat: number,
    hemat: number
}

export const ZERO_COST: Cost = {
    bmat: 0,
    rmat: 0,
    emat: 0,
    hemat: 0
}

export function calculateQueueCost(costPerCrate: number, crateCount: number) {
    let cost = 0;
    for (let i = 1; i <= crateCount; i++) {
        cost += Math.floor(costPerCrate * (1 - Math.min(i * 0.1, 0.5)))
    }
    return cost
}

export function add(a: Cost, b: Cost): Cost {
    return {
        bmat: a.bmat + b.bmat,
        emat: a.emat + b.emat,
        hemat: a.hemat + b.hemat,
        rmat: a.rmat + b.rmat
    }
}

export function calculateItemQueueCost(item: Item): Cost {
    if (item == undefined) return ZERO_COST
    const crateCount = CratePerQueue[item.itemCategory]
    return {
        bmat: calculateQueueCost(item.cost.bmat, crateCount),
        emat: calculateQueueCost(item.cost.emat, crateCount),
        hemat: calculateQueueCost(item.cost.hemat, crateCount),
        rmat: calculateQueueCost(item.cost.rmat, crateCount)
    }
}

export function asCrates(cost: Cost): Cost {
    return {
        bmat: Math.ceil(cost.bmat/ResourcePerCrate["bmat"]),
        emat: Math.ceil(cost.emat/ResourcePerCrate["emat"]),
        hemat: Math.ceil(cost.hemat/ResourcePerCrate["hemat"]),
        rmat: Math.ceil(cost.rmat/ResourcePerCrate["rmat"])
    }
}

File: calculator/src/commonMain/kotlin/de/chasenet/foxhole/model/Cost.kt
package de.chasenet.foxhole.model

import kotlinx.serialization.Serializable
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.min

@Serializable
data class Cost(
    val bmat: Int = 0,
    val rmat: Int = 0,
    val emat: Int = 0,
    val hemat: Int = 0,
) {
    operator fun plus(cost: Cost): Cost =
        Cost(
            bmat = bmat + cost.bmat,
            rmat = rmat + cost.rmat,
            emat = emat + cost.emat,
            hemat = hemat + cost.hemat,
        )

    operator fun get(resource: Resource): Int =
        when (resource) {
            Resource.bmat -> bmat
            Resource.rmat -> rmat
            Resource.emat -> emat
            Resource.hemat -> hemat
        }

    fun asCrates() =
        Cost(
            bmat = ceil(bmat / ResourcesPerCrate[Resource.bmat]!!.toFloat()).toInt(),
            emat = ceil(bmat / ResourcesPerCrate[Resource.emat]!!.toFloat()).toInt(),
            hemat = ceil(bmat / ResourcesPerCrate[Resource.hemat]!!.toFloat()).toInt(),
            rmat = ceil(bmat / ResourcesPerCrate[Resource.rmat]!!.toFloat()).toInt(),
        )

    fun queueCost(cratesInQueue: Int) =
        Cost(
            bmat = calculateQueueCost(bmat, cratesInQueue),
            rmat = calculateQueueCost(rmat, cratesInQueue),
            emat = calculateQueueCost(emat, cratesInQueue),
            hemat = calculateQueueCost(hemat, cratesInQueue),
        )
}

private fun calculateQueueCost(
    cost: Int,
    cratesInQueue: Int,
) = (0..cratesInQueue)
    .reduce { acc, crate ->
        acc + floor(cost.toFloat() * (1 - min(crate * 0.1, 0.5))).toInt()
    }
